src/common/io/src/buffer/buffer_read_datetime_ext.rs

```diff
@@ -61,48 +70,211 @@ where R: BufferRead
-                .or_else(|_| tz.datetime_from_str(v, "%Y-%m-%dT%H:%M:%S"))
-                .map_err_to_code(ErrorCode::BadBytes, || {
-                    format!("Cannot parse value:{:?} to DateTime type", v)
-                })?
-        };
+        // convert zero date to `1970-01-01`
+        if v == "0000-00-00" {
+            v = "1970-01-01";
+        }
+        let d = v
+            .parse::<NaiveDate>()
+            .map_err_to_code(ErrorCode::BadBytes, || {
+                format!("Cannot parse value:{} to Date type", v)
+            })?;
+        let mut dt = tz.from_local_datetime(&d.and_hms(0, 0, 0)).unwrap();
 
-        self.ignore(|b| b == b'z' || b == b'Z')?;
-        let dt = if self.ignore_byte(b'.')? {
-            buf.clear();
+        buf.clear();
+        let less_1000 = |dt: DateTime<Tz>| {
+            // convert timestamp less than `1000-01-01 00:00:00` to `1000-01-01 00:00:00`
+            if dt.year() < 1000 {
+                Ok(tz.from_utc_datetime(&NaiveDate::from_ymd(1000, 1, 1).and_hms(0, 0, 0)))
+            } else {
+                Ok(dt)
+            }
+        };
+        if self.ignore(|b| b == b' ' || b == b'T').unwrap() {
+            // Time HH:mm:ss
+            let mut buf = Vec::with_capacity(2);
             let size = self.keep_read(&mut buf, |f| (b'0'..=b'9').contains(&f))?;
-            let scales: i64 = lexical_core::FromLexical::from_lexical(buf.as_slice()).unwrap();
+            if buf.is_empty() || size > 2 {
+                return Err(ErrorCode::BadBytes(
+                    "err with parse time part. Format like this:[03:00:00]",
+                ));
+            }
+            let hour = lexical_core::FromLexical::from_lexical(buf.as_slice());
+            match hour {
+                Ok(hour) => {
+                    if self.ignore_byte(b':')? {
+                        buf.clear();
+                        let size = self.keep_read(&mut buf, |f| (b'0'..=b'9').contains(&f))?;
+                        if size == 0 {
+                            return Err(ErrorCode::BadBytes(
+                                "err with parse time part. Format like this:[03:00:00]",
+                            ));
+                        }
+                        let min: u32 =
+                            lexical_core::FromLexical::from_lexical(buf.as_slice()).unwrap();
+                        if size > 0 && size < 3 && self.ignore_byte(b':')? {
+                            buf.clear();
+                            let size = self.keep_read(&mut buf, |f| (b'0'..=b'9').contains(&f))?;
+                            if size == 0 {
+                                return Err(ErrorCode::BadBytes(
+                                    "err with parse time part. Format like this:[03:00:00]",
+                                ));
+                            }
+                            let sec: u32 =
+                                lexical_core::FromLexical::from_lexical(buf.as_slice()).unwrap();
```

Comment:
`hour`, `min`, and `sec` have the same structure, it looks like there is some duplicate code, maybe we can handle it with a loop, or write a function.

Reply:
I think we can not use loop. Because I need get hour , min and sec. These are different variable.

And if write a closer function, can not add self.xx to the closer func.

I try to write a closer func but need more match arm to avoid some err format.

```rust
let get_time = |buf: &mut Vec<u8>, size| {
                if size > 0 && size < 3 {
                    Ok(lexical_core::FromLexical::from_lexical(buf.as_slice()).unwrap())
                } else {
                    Err(ErrorCode::BadBytes(
                        "err with parse time part. Format like this:[03:00:00]",
                    ))
                }
            };
            let mut buf = Vec::with_capacity(2);
            let size = self.keep_read(&mut buf, |f| (b'0'..=b'9').contains(&f))?;
            match get_time(&mut buf, size) {
                Ok(hour) => {
                    if self.ignore_byte(b':')? {
                        buf.clear();
                        let size = self.keep_read(&mut buf, |f| (b'0'..=b'9').contains(&f))?;
                        match get_time(&mut buf, size) {
                            Ok(min) => {
                                if self.ignore_byte(b':')? {
                                    buf.clear();
                                    let size = self.keep_read(&mut buf, |f| (b'0'..=b'9').contains(&f))?;
                                    match get_time(&mut buf, size) {
                                        Ok(sec) => dt = tz.from_local_datetime(&d.and_hms(hour, min, sec)).unwrap(),
                                        Err(e) => return Err(e)
                                    }
                                } else {
                                    return Err(ErrorCode::BadBytes(
                                        "err with parse time part. Format like this:[03:00:00]",
                                    ));
                                }
                            },
                            Err(e) => return Err(e)
                        }
                    } else {
                        return Err(ErrorCode::BadBytes(
                            "err with parse time part. Format like this:[03:00:00]",
                        ));
                    }
                },
                Err(e) => return Err(e)
            }

```

Reply:
how about put `hour`, `min`, and `sec` in a vector? like this
```rust
let mut times Vec::with_capacity(3);
loop {
    let size = self.keep_read(&mut buf, |f| (b'0'..=b'9').contains(&f))?;
    match get_time(&mut buf, size) {
        Ok(time) => times.push(time),
        Err(e) => return e,
    }
    if times.len() == 3 {
        break;
    }
    self.must_ignore_byte(b':')?;
}
```

Reply:
That's a pretty good idea. I will do some test.